src/box/lua/net_box.lua

```diff
@@ -389,6 +389,13 @@ local function new_sm(uri, opts)
     local function weak_callback(...)
         local callback = weak_refs.callback
         if callback then return callback(...) end
+        -- The callback is responsible for handling graceful shutdown.
+        -- If it's garbage collected, the connection won't be closed on
+        -- receiving 'box.shutdown' event and so the server won't exit
+        -- until the connection object is garbage collected, which may
+        -- take forever. To avoid that, let's break the worker loop if
+        -- we see that the callback is unavailable.
+        error('garbage collected')
```

Comment:
That does not really look like an error. The owner was garbage collected like it is supposed to, correct? If you throw an error here, then you will get it in the logs which might be confusing and even look like a bug (some strange message `"garbage collected"` coming from an internal file `net_box.lua`). The logging will happen in `netbox_transport_on_state_change_pcall()`.

I would suggest to find a way to break the loop without raising an error.

Reply:
Right. I didn't think about error logging. Replaced with `transport:stop()`. Thanks for catching this.